cudamapper/src/cudamapper_utils.cpp

```diff
@@ -8,62 +8,143 @@
+                                                 overlaps[i].query_start_position_in_read_,
+                                                 overlaps[i].query_end_position_in_read_,
+                                                 static_cast<unsigned char>(overlaps[i].relative_strand),
+                                                 target_read_name.c_str(),
+                                                 target_parser.get_sequence_by_id(overlaps[i].target_read_id_).seq.length(),
+                                                 overlaps[i].target_start_position_in_read_,
+                                                 overlaps[i].target_end_position_in_read_,
+                                                 overlaps[i].num_residues_ * kmer_size, // Print out the number of residue matches multiplied by kmer size to get approximate number of matching bases
+                                                 std::max(std::abs(static_cast<std::int64_t>(overlaps[i].target_start_position_in_read_) - static_cast<std::int64_t>(overlaps[i].target_end_position_in_read_)),
+                                                          std::abs(static_cast<std::int64_t>(overlaps[i].query_start_position_in_read_) - static_cast<std::int64_t>(overlaps[i].query_end_position_in_read_))), //Approximate alignment length
+                                                 255);
+        chars_in_buffer += added_chars;
+        // If CIGAR string is generated, output in PAF.
+        if (!cigar.empty())
+        {
+            const int32_t added_cigar_chars = std::sprintf(buffer.data() + chars_in_buffer,
+                                                           "\tcg:Z:%s",
+                                                           cigar[i].c_str());
+            chars_in_buffer += added_cigar_chars;
+        }
+        // Add new line to demarcate new entry.
+        buffer.data()[chars_in_buffer] = '\n';
+        ++chars_in_buffer;
+    }
+    buffer[chars_in_buffer] = '\0';
+
+    std::lock_guard<std::mutex> lg(write_output_mutex);
+    printf("%s", buffer.data());
+}
+
+} // namespace
+
+void print_paf(const std::vector<Overlap>& overlaps,
+               const std::vector<std::string>& cigar,
+               const io::FastaParser& query_parser,
+               const io::FastaParser& target_parser,
+               const int32_t kmer_size,
+               std::mutex& write_output_mutex,
+               const int32_t number_of_devices)
+{
+    assert(!cigar.empty() || (overlaps.size() == cigar.size()));
+
+    // divide the work into several threads
+
+    int32_t number_of_threads   = std::thread::hardware_concurrency() / number_of_devices; // We could use a better heuristic here
+    int64_t overlaps_per_thread = get_size<int64_t>(overlaps) / number_of_threads;
+
+    std::vector<std::thread> threads;
+
+    for (uint32_t thread_id = 0; thread_id < number_of_threads; ++thread_id)
```

Comment:
Optional: Why `uint32_t` here all of the sudden?

Reply:
Again, initially I was using `uint32_t`, then I switched to `int32_t` and forgot to change
Done